wgengine/magicsock/peermtu.go

```diff
@@ -0,0 +1,89 @@
+
+package magicsock
+
+// Peer path MTU routines shared by platforms that implement it.
+
+// DontFragSetting returns true if at least one of the underlying sockets of
+// this connection is a UDP socket with the don't fragment bit set, otherwise it
+// returns false.
+func (c *Conn) DontFragSetting() bool {
+	return c.getDontFragment("udp4") || c.getDontFragment("udp6")
+}
+
+// ShouldPMTUD returns true if this client should try to enable peer MTU
+// discovery, false otherwise.
+func (c *Conn) ShouldPMTUD() bool {
+	if v, ok := debugEnablePMTUD().Get(); ok {
+		if debugPMTUD() {
+			c.logf("magicsock: peer path MTU discovery set via envknob to %v", v)
+		}
+		return v
+	}
+	if c.controlKnobs != nil {
+		if v, ok := c.controlKnobs.PeerMTUEnable.Load().Get(); ok {
+			if debugPMTUD() {
+				c.logf("magicsock: peer path MTU discovery set by control to %v", v)
+			}
+			return v
+		}
+	}
+	if debugPMTUD() {
+		c.logf("magicsock: peer path MTU discovery set by default to false")
+	}
+	return false // Until we feel confident PMTUD is solid.
+}
+
+// PeerMTUEnabled returns true if this Conn is has peer path MTU discovery enabled.
+func (c *Conn) PeerMTUEnabled() bool {
+	return c.peerMTUEnabled.Load()
+}
+
+// UpdatePMTUD configures underlying sockets of this Conn to enable or disable
+// peer path MTU discovery according to the current configuration.
+//
+// Enabling or disabling peer path MTU discovery requires setting the don't
+// fragment bit on its two underlying pconns. There are three distinct results
+// for this operation on each pconn:
+//
+// 1. Success
+// 2. Failure (not supported on this platform, or supported but failed)
+// 3. Not a UDP socket (most likely one of IPv4 or IPv6 couldn't be used)
+//
+// To simplify the fast path for the most common case, we set the PMTUD status
+// of the overall Conn according to the results of setting the sockopt on pconn
+// as follows:
+//
+// 1. Both setsockopts succeed: PMTUD status update succeeds
+// 2. One succeeds, one returns not a UDP socket: PMTUD status update succeeds
+// 4. Neither setsockopt succeeds: PMTUD disabled
+// 3. Either setsockopt fails: PMTUD disabled
+func (c *Conn) UpdatePMTUD() {
+	if debugPMTUD() {
+		c.logf("magicsock: peer MTU status %v DF bit status: v4: %v v6: %v", c.peerMTUEnabled.Load(), c.getDontFragment("udp4"), c.getDontFragment("udp6"))
+	}
+
+	enable := c.ShouldPMTUD()
+	if c.peerMTUEnabled.Load() == enable {
+		c.logf("magicsock: peer MTU status is %v", enable)
+		return
+	}
+	newStatus := enable
+	err4 := c.setDontFragment("udp4", enable)
+	err6 := c.setDontFragment("udp6", enable)
+	anySuccess := err4 == nil || err6 == nil
+	noFailures := (err4 == nil || err4 == errNoActiveUDP) && (err6 == nil || err6 == errNoActiveUDP)
+
+	if anySuccess && noFailures {
+		c.logf("magicsock: peer MTU status updated to %v", newStatus)
+	} else {
+		c.logf("magicsock: updating peer MTU status to %v failed (v4: %v, v6: %v), disabling", enable, err4, err6)
```

Comment:
Is it possible that we take this branch because we happen to be mid rebind? If so do we end up permanently applying this state?

Reply:
Good question! The answer is no since this is only called in two places: initialization, and in reconfig() under wg lock, and rebind has just happened (if it is going to happen at all). I was wondering how to document this since the rest of the things that happen in reconfig() don't have any obvious concurrency-related documentation that I can see?

The actual theory for failure of the setsockopt is ???. I don't know a thing that will cause it to fail that is not a bug in our code (which I did find one already).

On state: we will keep trying to implement the result of ShouldPMTUD() every time we do a reconfig. That's part of why I took the setsockopt out of the rebind() code, since if there was some persistent error we would rebind on every reconfig.

I would love any thoughts on how to best document this stuff!